src/java.base/share/classes/java/lang/ClassLoader.java

```diff
@@ -753,6 +761,86 @@ protected Class<?> findClass(String moduleName, String name) {
         return null;
     }
 
+    // The threadLoadingClassMap is a concurrent hashtable that keeps track of the threads
+    // that are currently loading each class by this class loader.  The class name is mapped to
+    // the thread that is first seen loading that class, then removed when loading is complete
+    // for that class or throws an exception.
+    private final Thread curThreadLoadingClass(String name, Thread thread) {
+        return threadLoadingClassMap.putIfAbsent(name, thread);
+    }
+
+    private final void removeThreadLoadingClass(String name) {
+        threadLoadingClassMap.remove(name);
+        // Notify threads waiting on the class loader lock
+        // that this class has been loaded or failed.
+        notifyAll();
+    }
+
+    // We only get here if the application has released the
+    // classloader lock (wait) when another thread was in the middle of loading a
+    // superclass/superinterface for this class, and now
+    // this thread is also trying to load this class.
+    // To minimize surprises, this thread waits while the first thread
+    // that started to load a class completes the loading or fails.
+    @SuppressWarnings("removal")
+    private Thread waitForThreadLoadingClass(String name, Thread currentThread) {
+        Thread thread;
+        boolean interrupted = false;
+        while ((thread = curThreadLoadingClass(name, currentThread)) != null) {
```

Comment:
Should this be:

```suggestion
        while ((thread = curThreadLoadingClass(name, currentThread)) != null && thread != currentThread) {
```

This method should return either no thread is loading the named class or it's currentThread which means circularity.

Reply:
Is circularity even possible?

Reply:
Yes it is possible.  So we want to return if the thread returned isn't the current thread (not wait) because we have to throw CCE from there.  This case is found running Xcomp.  This is the equivalent case for the null class loader:
    // -Xcomp calls load_signature_classes which might result in loading
    // a class that's already in the process of loading, so we detect CCE here also.